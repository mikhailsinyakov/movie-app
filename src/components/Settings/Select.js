import React, { useState, useEffect, useContext } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { HomeStateContext } from "components/HomeState";

const Select = ({ options, initValue, handleChange, showAlways, className }) => {
  const { isSearchActive } = useContext(HomeStateContext);
  const [value, setValue] = useState(initValue);
  const [show, setShow] = useState(!isSearchActive);
  
  useEffect(() => {
    if (!showAlways) {
      if (!isSearchActive) {
        setTimeout(() => {
          setShow(true);
        }, 500);
      } else setShow(false);
    }
  }, [showAlways, isSearchActive]);
  
  const style = {
    display: showAlways ? "inline" : show ? "inline" : "none"
  };

  const onChange = e => {
    setValue(e.target.value);
    handleChange(e.target.value);
  };

  return (
    <select 
      value={value} 
      onChange={onChange} 
      className={className}
      style={style}
    >
      {options.map(({ value, name }) => (
        <option key={value} value={value}>
          {name}
        </option>
      ))}
    </select>
  );
};

Select.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ),
  initValue: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  showAlways: PropTypes.bool.isRequired
};



const StyledSelect = styled(Select)`
  background: #e6e6e8;
  height: 30px;
  border: 2px solid #739690;
  border-radius: 0.4rem;
  padding: 0 0.5rem;
  font-family: inherit;
  margin: 0.5rem 0.5rem;

  &:focus {
    outline: none;
  }
`;

export default StyledSelect;
