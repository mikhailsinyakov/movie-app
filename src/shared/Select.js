import React, { useState } from "react";
import PropTypes from "prop-types";

const Select = ({ options, initValue, handleChange }) => {
  const [value, setValue] = useState(initValue);

  const onChange = e => {
    setValue(e.target.value);
    handleChange(e.target.value);
  };

  return (
    <select value={value} onChange={onChange}>
      {options.map(({ value, name }) => (
        <option key={value} value={value}>
          {name}
        </option>
      ))}
    </select>
  );
};

Select.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ),
  initValue: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired
};

export default Select;
