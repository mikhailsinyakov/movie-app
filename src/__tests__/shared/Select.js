import React from "react";
import Renderer from "react-test-renderer";
import { screen, fireEvent, render } from "@testing-library/react";
import Select from "../../shared/Select";

const options = [
  { value: "value1", name: "name1" },
  { value: "value2", name: "name2" }
];

const initValue = "value1";

const testRenderer = Renderer.create(
  <Select options={options} initValue={initValue} handleChange={() => {}} />
);

it("matches snapshot", () => {
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

it("call handleChange function when clicking on different option", () => {
  const handleChangeMock = jest.fn();

  const { container } = render(
    <Select
      options={options}
      initValue={initValue}
      handleChange={handleChangeMock}
    />
  );

  fireEvent.change(container.getElementsByTagName("select")[0], {
    target: { value: "value2" }
  });
  expect(handleChangeMock).toHaveBeenCalledWith("value2");
});
