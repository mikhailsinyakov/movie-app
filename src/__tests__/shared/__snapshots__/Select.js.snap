// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<select
  onChange={[Function]}
  value="value1"
>
  <option
    value="value1"
  >
    name1
  </option>
  <option
    value="value2"
  >
    name2
  </option>
</select>
`;
